File: Dynamic Array/code/Vector.java
/* 
 * @author: Ahmed Elshamy
 *
 * This program demonstrates the implementation of a dynamic vector with various operations
 * such as adding, removing, rotating, and finding elements. The Vector class allows flexibility
 * with dynamic resizing and efficient access to vector elements.
*/

import java.util.Arrays;

public class Vector {
    // Internal array to store elements
    private int[] arr;
    // Current number of elements in the vector
    private int size;
    // Total capacity of the internal array
    private int capacity;

    /**
     * Constructor to initialize vector with given size
     * If size is invalid (<=0), defaults to 1
     */
    public Vector(int si) {
        if (si <= 0) {
            si = 1;
        }
        size = si;
        capacity = size * 2;
        arr = new int[capacity];
    }

    /**
     * Copy constructor - creates a new vector as a copy of another
     */
    public Vector(Vector other) {
        this.size = other.size;
        this.capacity = other.capacity;
        this.arr = Arrays.copyOf(other.arr, capacity);
    }

    /**
     * Returns the first element of the vector
     * Throws exception if vector is empty
     */
    public int front() {
        if (size == 0)
            throw new RuntimeException("Vector is empty");
        return arr[0];
    }

    /**
     * Returns the last element of the vector
     * Throws exception if vector is empty
     */
    public int back() {
        if (size == 0)
            throw new RuntimeException("Vector is empty");
        return arr[size - 1];
    }

    /**
     * Sets the value at specified index
     * Throws exception if index is out of range
     */
    public void set(int index, int val) {
        if (index < 0 || index >= size)
            throw new IndexOutOfBoundsException("Index out of range");
        arr[index] = val;
    }

    /**
     * Adds a new element to the end of the vector
     * Expands capacity if necessary
     */
    public void pushBack(int val) {
        if (size == capacity) {
            expandCapacity();
        }
        arr[size++] = val;
    }

    /**
     * Inserts an element at specified index
     * Shifts elements to make space
     * Throws exception if index is out of range
     */
    public void insert(int index, int val) {
        if (index < 0 || index > size)
            throw new IndexOutOfBoundsException("Index out of range");
        if (size == capacity) {
            expandCapacity();
        }
        for (int i = size; i > index; i--) {
            arr[i] = arr[i - 1];
        }
        arr[index] = val;
        size++;
    }

    /**
     * Returns the current size of the vector
     */
    public int getSize() {
        return size;
    }

    /**
     * Returns the element at specified index
     * Throws exception if index is out of range
     */
    public int get(int index) {
        if (index < 0 || index >= size)
            throw new IndexOutOfBoundsException("Index out of range");
        return arr[index];
    }

    /**
     * Finds the first occurrence of a value
     * Returns the index if found, -1 otherwise
     */
    public int find(int val) {
        for (int i = 0; i < size; i++) {
            if (arr[i] == val)
                return i;
        }
        return -1;
    }

    /**
     * Prints all elements of the vector
     */
    public void print() {
        for (int i = 0; i < size; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    /**
     * Rotates all elements one position to the right
     * Last element becomes the first
     */
    public void rightRotate() {
        if (size <= 1)
            return;
        int last = arr[size - 1];
        for (int i = size - 1; i > 0; i--) {
            arr[i] = arr[i - 1];
        }
        arr[0] = last;
    }

    /**
     * Rotates elements to the right by specified number of positions
     * Uses the three-reverse method for efficiency
     */
    public void rightRotate(int times) {
        if (times <= 0 || size <= 1)
            return;
        times %= size;
        reverse(0, size - 1);
        reverse(0, times - 1);
        reverse(times, size - 1);
    }

    /**
     * Rotates all elements one position to the left
     * First element becomes the last
     */
    public void leftRotate() {
        if (size <= 1)
            return;
        int first = arr[0];
        for (int i = 0; i < size - 1; i++) {
            arr[i] = arr[i + 1];
        }
        arr[size - 1] = first;
    }

    /**
     * Removes and returns the element at specified index
     * Shifts elements to fill the gap
     * Throws exception if index is out of range
     */
    public int pop(int index) {
        if (index < 0 || index >= size)
            throw new IndexOutOfBoundsException("Index out of range");
        int element = arr[index];
        for (int i = index; i < size - 1; i++) {
            arr[i] = arr[i + 1];
        }
        size--;
        return element;
    }

    /**
     * Finds a value and moves it one position closer to the front (transposition)
     * Used as a heuristic to improve future search performance
     * Returns the new index if found, -1 otherwise
     */
    public int findTransposition(int val) {
        for (int i = 0; i < size; i++) {
            if (arr[i] == val) {
                if (i == 0)
                    return i;
                int temp = arr[i];
                arr[i] = arr[i - 1];
                arr[i - 1] = temp;
                return i - 1;
            }
        }
        return -1;
    }

    /**
     * Helper method to reverse a portion of the array
     * Used by rightRotate method
     */
    private void reverse(int start, int end) {
        while (start < end) {
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }
    }

    /**
     * Doubles the capacity of the internal array
     * Used when more space is needed
     */
    private void expandCapacity() {
        capacity = capacity == 0 ? 1 : capacity * 2;
        arr = Arrays.copyOf(arr, capacity);
    }

    public static void main(String[] args) {
        Vector v = new Vector(5);
        System.out.print("Initial vector (size 5): ");
        v.print();

        v.pushBack(10);
        v.pushBack(20);
        v.pushBack(30);
        System.out.print("After adding elements (10, 20, 30): ");
        v.print();

        System.out.println("First element (front): " + v.front());
        System.out.println("Last element (back): " + v.back());
        v.set(0, 99);
        System.out.print("After setting the first element to 99: ");
        v.print();

        v.insert(2, 50);
        System.out.print("After inserting 50 at index 2: ");
        v.print();

        int index = v.find(20);
        if (index != -1)
            System.out.println("Element 20 found at index: " + index);
        else
            System.out.println("Element 20 not found");

        v.rightRotate();
        System.out.print("After right rotation by 1: ");
        v.print();

        v.rightRotate(3);
        System.out.print("After right rotation by 3: ");
        v.print();

        v.leftRotate();
        System.out.print("After left rotation by 1: ");
        v.print();

        int removed = v.pop(2);
        System.out.println("After removing element at index 2 (removed: " + removed + "): ");
        v.print();

        index = v.findTransposition(30);
        if (index != -1)
            System.out.println("Element 30 found and moved closer to the front. New index: " + index);
        else
            System.out.println("Element 30 not found");
        System.out.print("After find with transposition: ");
        v.print();

        try {
            System.out.println("Accessing element at invalid index (-1): " + v.get(-1));
        } catch (IndexOutOfBoundsException e) {
            System.out.println("Caught exception: " + e.getMessage());
        }

        try {
            System.out.print("Popping element at invalid index (100): ");
            v.pop(100);
        } catch (IndexOutOfBoundsException e) {
            System.out.println("Caught exception: " + e.getMessage());
        }

        System.out.print("Final state of the vector: ");
        v.print();
    }
}

File: Dynamic Array/code/Vector.cpp
/**
 * @author: Ahmed Elshamy
 * @time: 18/04/2025  00:55 PM
 *
 * This program demonstrates the implementation of a dynamic vector with various operations
 * such as adding, removing, rotating, and finding elements. The Vector class allows flexibility
 * with dynamic resizing and efficient access to vector elements.
 */

#include <iostream>
#include <cassert>
#include <stdexcept> // For exception handling
#include <algorithm> // For std::reverse
using namespace std;

class Vector
{
private:
    int *arr = nullptr; // Pointer to the array holding vector elements
    int size = 0;       // Number of elements currently in the vector
    int capacity = 0;   // Capacity of the vector (maximum elements it can hold)

    // Expands the vector capacity when it's full
    void expand_capacity();

public:
    // Constructor to initialize the vector with a given size
    Vector(int si);
    // Copy constructor for deep copying
    Vector(const Vector &other);
    // Assignment operator for deep copying
    Vector &operator=(const Vector &other);
    // Destructor to free the dynamically allocated memory
    ~Vector();

    int front() const;               // Returns the first element of the vector
    int back() const;                // Returns the last element of the vector
    void set(int index, int val);    // Sets the element at a given index
    void push_back(int val);         // Adds a new element to the end of the vector
    void insert(int index, int val); // Inserts an element at a given index
    int get_size() const;            // Returns the current size of the vector
    int get(int index) const;        // Returns the element at a given index
    void print() const;              // Prints all elements in the vector
    int find(int val) const;         // Returns the index of the first occurrence of a value
    void right_rotate();             // Rotates the vector to the right by 1 position
    void right_rotate(int times);    // Rotates the vector to the right by 'times' positions
    void left_rotate();              // Rotates the vector to the left by 1 position
    int pop(int index);              // Removes and returns the element at the given index
    int find_transposition(int val); // Finds and moves a value closer to the front
};

// Expands the vector's capacity by doubling its current size
void Vector::expand_capacity()
{
    capacity = capacity == 0 ? 1 : capacity * 2; // Double capacity (or set to 1 if 0)
    int *temp_arr = new int[capacity]{};         // Allocate a new array with expanded capacity

    // Copy old array into the new array
    for (int i = 0; i < size; ++i)
    {
        temp_arr[i] = arr[i];
    }

    swap(arr, temp_arr); // Swap old array with the new one
    delete[] temp_arr;   // Free the old array
}

// Constructor to initialize vector with a specified size
Vector::Vector(int si)
{
    if (si <= 0)
    {
        si = 1; // Ensure size is at least 1
    }
    size = si;
    capacity = size * 2;       // Initial capacity is twice the size
    arr = new int[capacity]{}; // Allocate memory for the vector
}

// Copy constructor for deep copying
Vector::Vector(const Vector &other)
{
    size = other.size;
    capacity = other.capacity;
    arr = new int[capacity];
    for (int i = 0; i < size; ++i)
    {
        arr[i] = other.arr[i];
    }
}

// Assignment operator for deep copying
Vector &Vector::operator=(const Vector &other)
{
    if (this == &other)
        return *this;

    delete[] arr;

    size = other.size;
    capacity = other.capacity;
    arr = new int[capacity];
    for (int i = 0; i < size; ++i)
    {
        arr[i] = other.arr[i];
    }

    return *this;
}

// Destructor to free the dynamically allocated memory
Vector::~Vector()
{
    delete[] arr;  // Deallocate the array memory
    arr = nullptr; // Prevent dangling pointer
    size = 0;
    capacity = 0;
}

// Returns the first element of the vector
int Vector::front() const
{
    if (size == 0)
        throw runtime_error("Vector is empty");
    return arr[0];
}

// Returns the last element of the vector
int Vector::back() const
{
    if (size == 0)
        throw runtime_error("Vector is empty");
    return arr[size - 1];
}

// Sets the element at the given index
void Vector::set(int index, int val)
{
    if (index < 0 || index >= size)
        throw out_of_range("Index out of range");
    arr[index] = val;
}

// Adds a new element to the end of the vector
void Vector::push_back(int val)
{
    if (size == capacity)
    {
        expand_capacity(); // Expand capacity if full
    }
    arr[size++] = val;
}

// Returns the element at a given index
int Vector::get(int index) const
{
    if (index < 0 || index >= size)
        throw out_of_range("Index out of range");
    return arr[index];
}

// Finds the index of the first occurrence of the given value
int Vector::find(int val) const
{
    for (int i = 0; i < size; ++i)
    {
        if (arr[i] == val)
        {
            return i; // Value found, return its index
        }
    }
    return -1; // Return -1 if value is not found
}

// Inserts an element at the given index
void Vector::insert(int index, int val)
{
    if (index < 0 || index > size)
        throw out_of_range("Index out of range");
    if (size == capacity)
    {
        expand_capacity(); // Expand capacity if necessary
    }
    // Shift elements to the right to make space for the new element
    for (int i = size; i > index; --i)
    {
        arr[i] = arr[i - 1];
    }
    arr[index] = val; // Insert the new element
    ++size;
}

// Prints all elements of the vector
void Vector::print() const
{
    for (int i = 0; i < size; ++i)
    {
        cout << arr[i] << " ";
    }
    cout << '\n';
}

// Rotates the vector to the right by one position
void Vector::right_rotate()
{
    if (size <= 1)
        return;

    int last = arr[size - 1];
    for (int i = size - 1; i > 0; --i)
    {
        arr[i] = arr[i - 1]; // Shift elements to the right
    }
    arr[0] = last; // Move the last element to the front
}

// Rotates the vector to the right by a given number of positions
void Vector::right_rotate(int times)
{
    if (times <= 0 || size <= 1)
        return; // No need to rotate if times is 0 or size is 1

    times %= size; // Normalize the number of rotations

    // Reverse the entire array
    reverse(arr, arr + size);
    // Reverse the first 'times' elements
    reverse(arr, arr + times);
    // Reverse the remaining elements
    reverse(arr + times, arr + size);
}

// Rotates the vector to the left by one position
void Vector::left_rotate()
{
    if (size <= 1)
        return;

    int first = arr[0];
    for (int i = 0; i < size - 1; ++i)
    {
        arr[i] = arr[i + 1]; // Shift elements to the left
    }
    arr[size - 1] = first; // Move the first element to the end
}

// Removes and returns the element at the given index
int Vector::pop(int index)
{
    if (index < 0 || index >= size)
        throw out_of_range("Index out of range");

    int element = arr[index]; // Store the element to return
    for (int i = index; i < size - 1; ++i)
    {
        arr[i] = arr[i + 1]; // Shift elements to the left
    }
    --size; // Decrease the size of the vector
    return element;
}

// Finds the value and moves it closer to the front (transposition)
int Vector::find_transposition(int val)
{
    for (int i = 0; i < size; ++i)
    {
        if (arr[i] == val)
        {
            if (i == 0)
                return i;             // If it's already at the front, return the index
            swap(arr[i], arr[i - 1]); // Swap with the previous element
            return i - 1;             // Return the new index
        }
    }
    return -1; // Return -1 if the value is not found
}

// Main function to test the Vector class
int main()
{
    // Create a vector of size 5 and initialize it
    Vector v(5);
    cout << "Initial vector (size 5): ";
    v.print();

    // Add elements to the vector
    v.push_back(10);
    v.push_back(20);
    v.push_back(30);
    cout << "After adding elements (10, 20, 30): ";
    v.print();

    // Access and modify elements
    cout << "First element (front): " << v.front() << '\n';
    cout << "Last element (back): " << v.back() << '\n';
    v.set(0, 99); // Modify the first element
    cout << "After setting the first element to 99: ";
    v.print();

    // Insert an element at a specific index
    v.insert(2, 50); // Insert 50 at index 2
    cout << "After inserting 50 at index 2: ";
    v.print();

    // Find an element
    int index = v.find(20);
    if (index != -1)
        cout << "Element 20 found at index: " << index << '\n';
    else
        cout << "Element 20 not found\n";

    // Test right rotation by 1
    v.right_rotate();
    cout << "After right rotation by 1: ";
    v.print();

    // Test right rotation by multiple positions
    v.right_rotate(3);
    cout << "After right rotation by 3: ";
    v.print();

    // Test left rotation
    v.left_rotate();
    cout << "After left rotation by 1: ";
    v.print();

    // Remove an element at a specific index
    int removed = v.pop(2); // Remove the element at index 2
    cout << "After removing element at index 2 (removed: " << removed << "): ";
    v.print();

    // Test find with transposition
    index = v.find_transposition(30);
    if (index != -1)
        cout << "Element 30 found and moved closer to the front. New index: " << index << '\n';
    else
        cout << "Element 30 not found\n";
    cout << "After find with transposition: ";
    v.print();

    // Test edge cases
    try
    {
        cout << "Accessing element at invalid index (-1): ";
        cout << v.get(-1) << '\n';
    }
    catch (const out_of_range &e)
    {
        cout << "Caught exception: " << e.what() << '\n';
    }

    try
    {
        cout << "Popping element at invalid index (100): ";
        v.pop(100);
    }
    catch (const out_of_range &e)
    {
        cout << "Caught exception: " << e.what() << '\n';
    }

    // Final state of the vector
    cout << "Final state of the vector: ";
    v.print();

    return 0;
}
